File: Graph&Maze/200-NumberOfIslands.java
class Solution {
    public void dfs(char[][] grid, int r, int c){
        int row = grid.length;
        int col = grid[0].length;
        if (r<0 || c<0 || r > row-1 || c>col-1 || grid[r][c]=='0'){
            return;
        }

        grid[r][c]='0';
        dfs(grid, r-1, c);
        dfs(grid, r+1, c);
        dfs(grid, r, c+1);
        dfs(grid, r, c-1);
    }

    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int nr = grid.length;
        int nc = grid[0].length;
        int num_islands = 0;

        for (int r = 0; r < nr; ++r) {
            for (int c = 0; c < nc; ++c) {
                if (grid[r][c] == '1') {
                    dfs(grid, r, c);
                    num_islands++;
                }
            }
        }
        return num_islands;
    }
}

File: python_july/200. Number of Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        def dfs(row, col):
            if row < 0 or col < 0 or row >=m or col >= n or grid[row][col] != '1':
                return 
            grid[row][col] = '0'
            
            dfs(row-1, col)
            dfs(row+1, col)
            dfs(row, col-1)
            dfs(row, col+1)
        
        cnt = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    dfs(i, j)
                    cnt +=1
        return cnt
        